feat: ADR-040: Implement BadgerDB backend

roysc opened the pull request:
## Description

Partially resolves: https://github.com/vulcanize/cosmos-sdk/issues/14

Implements a [BadgerDB](https://pkg.go.dev/github.com/dgraph-io/badger/v3)-based backend for the DB interface introduced by https://github.com/cosmos/cosmos-sdk/pull/9573 and specified by [ADR-040](https://github.com/cosmos/cosmos-sdk/blob/eb7d939f86c6cd7b4218492364cdda3f649f06b5/docs/architecture/adr-040-storage-and-smt-state-commitments.md). This uses Badger's "managed" mode for version management, and supports optimistically concurrent transactions (required one [patch](https://github.com/dgraph-io/badger/pull/1716) upstream to fix handling of write conflicts).

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules) - n/a
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

roysc commented:
As this depends on https://github.com/cosmos/cosmos-sdk/pull/9573, I'm making it a draft for now and will rebase when that is merged.

Also, not sure if more documentation is needed beyond godoc comments.

alexanderbez reviewed (commented):
I'm going to wait until the mem-db PR is reviewed and merged before reviewing RocksDB and BadgerDB variants. We can rebase on `master` after it's merged.

marbar3778 commented:
@roysc can this be rebased. Then we can begin reviewing

roysc commented:
Rebased and ready for review

alexanderbez commented on db/README.md:
```diff
@@ -37,3 +37,28 @@ This represents a self-contained and immutable view of a database's version hist
 The in-memory DB in the `db/memdb` package cannot be persisted to disk. It is implemented using the Google [btree](https://pkg.go.dev/github.com/google/btree) library.
   * This currently does not perform write conflict detection, so it only supports a single open write-transaction at a time. Multiple and concurrent read-transactions are supported.
+
+### BadgerDB ###
```
```suggestion
### BadgerDB
```

alexanderbez commented on db/README.md:
```diff
@@ -37,3 +37,28 @@ This represents a self-contained and immutable view of a database's version hist
+
+### BadgerDB ###
+
+A [BadgerDB](https://pkg.go.dev/github.com/dgraph-io/badger/v3)-based backend. Internally, this uses BadgerDB's "managed" mode for version management.
```
nit: can you link to the "managed" mode document/link please?

alexanderbez commented on db/badgerdb/db.go:
```diff
@@ -0,0 +1,402 @@
+	versionsFilename string = "versions.csv"
+)
+
+var _ dbm.DBConnection = (*BadgerDB)(nil)
+var _ dbm.DBReader = (*badgerTxn)(nil)
+var _ dbm.DBWriter = (*badgerWriter)(nil)
+var _ dbm.DBReadWriter = (*badgerWriter)(nil)
```
nit: group the vars :)

alexanderbez commented on db/badgerdb/db.go:
```diff
@@ -0,0 +1,402 @@
+	db   *badger.DB
+	vmgr *versionManager
+	mtx  sync.RWMutex
+	// Track open DBWriters
```
```suggestion
```

alexanderbez commented on db/badgerdb/db.go:
```diff
@@ -0,0 +1,402 @@
+
+type badgerTxn struct {
+	txn *badger.Txn
+	// vmgr *versionManager
```
Do we need this?

alexanderbez commented on db/badgerdb/db.go:
```diff
@@ -0,0 +1,402 @@
+	// vmgr *versionManager
+	db *BadgerDB
+}
+type badgerWriter struct {
```
```suggestion

type badgerWriter struct {
```

alexanderbez commented on db/badgerdb/db.go:
```diff
@@ -0,0 +1,402 @@
+	*dbm.VersionManager
+	vmap   versionTsMap
+	lastTs uint64
+}
```
```suggestion
}

```

alexanderbez commented on db/badgerdb/db.go:
```diff
@@ -0,0 +1,402 @@
+	vmap   versionTsMap
+	lastTs uint64
+}
+type versionTsMap map[uint64]uint64
```
nit: does this need to be a type alias?

roysc replied:
no, removed

alexanderbez commented on db/badgerdb/db.go:
```diff
@@ -0,0 +1,402 @@
+		return nil, err
+	}
+	opts := badger.DefaultOptions(dir)
+	// todo: NumVersionsToKeep
```
Please reference an issue for TODOs 🙏

alexanderbez commented on db/badgerdb/db.go:
```diff
@@ -0,0 +1,402 @@
+		versions = append(versions, version)
+		vmap[version] = ts
+	}
+	if ts == 0 {
+		ts = 1
+	}
```
Can you add a comment for this code block pls?

roysc replied:
Actually, this can be dropped. It was there to handle the empty case, and ensure the max timestamp is > 0, but it's not needed.
This made me notice a different bug, though - this code and the versionmanager constructor should not require sorted input. Will fix.

alexanderbez commented on db/badgerdb/db.go:
```diff
@@ -0,0 +1,402 @@
+	if err != nil {
+		return nil, err
+	}
+	var versions []uint64
+	var ts uint64
```
nit: group these vars :)

roysc commented:
Along with review follow-ups, I fixed the version manager initialization so it's not dependent on ordered input, and updated a test case to verify its behavior.

marbar3778 reviewed (approved):
utACK

alexanderbez commented:
We can merge this, yes?
